otherlibs/win32unix/createprocess.c

```diff
@@ -25,32 +25,50 @@
 static int win_has_console(void);
 
 static DWORD do_create_process_native(wchar_t * exefile, wchar_t * cmdline,
-                                      wchar_t * env, HANDLE fd1, HANDLE fd2,
-                                      HANDLE fd3, HANDLE * hProcess)
+                                      wchar_t * env, HANDLE fd0, HANDLE fd1,
+                                      HANDLE fd2, HANDLE * hProcess)
 {
   PROCESS_INFORMATION pi;
   STARTUPINFO si;
-  DWORD flags, err;
+  DWORD handleInfo[3], flags, err;
   HANDLE hp;
 
   err = ERROR_SUCCESS;
   /* Prepare stdin/stdout/stderr redirection */
   ZeroMemory(&si, sizeof(STARTUPINFO));
   si.cb = sizeof(STARTUPINFO);
   si.dwFlags = STARTF_USESTDHANDLES;
-  /* Duplicate the handles fd1, fd2, fd3 to make sure they are inheritable */
+  /* If needed, duplicate the handles fd0, fd1, fd2 to make sure they
+     are inheritable. */
+  if (! GetHandleInformation(fd0, &handleInfo[0]) ||
+      ! GetHandleInformation(fd1, &handleInfo[1]) ||
+      ! GetHandleInformation(fd2, &handleInfo[2])) {
+    goto ret1;
+  }
   hp = GetCurrentProcess();
-  if (! DuplicateHandle(hp, fd1, hp, &(si.hStdInput),
-                        0, TRUE, DUPLICATE_SAME_ACCESS)) {
-    err = GetLastError(); goto ret1;
+  if (! (handleInfo[0] & HANDLE_FLAG_INHERIT)) {
+    if (! DuplicateHandle(hp, fd0, hp, &(si.hStdInput),
+                          0, TRUE, DUPLICATE_SAME_ACCESS)) {
+      err = GetLastError(); goto ret1;
+    }
+  } else {
+      si.hStdInput = fd0;
```

Comment:
Could we factor out this logic in a function?  (Then we'd call the function 3 times for each of the 3 standard handles.)  Something like:
- GetHandleInformation
- if inheritable, return original handle
- otherwise, duplicate to an inheritable handle

Reply:
Done in the third commit. The `ensure_inheritable` function must also return the information that we've duplicated the handle to close it after the child process is spawned.